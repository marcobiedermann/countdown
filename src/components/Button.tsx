import clsx from 'clsx';
import type { ReactNode } from 'react';
import { Link } from 'react-router-dom';

interface BaseProps {
  children: ReactNode;
  style?: 'pill' | 'round';
  variant?: 'primary' | 'secondary';
}

interface ButtonLinkProps extends BaseProps {
  to: string;
}

interface ButtonProps extends BaseProps {
  type: 'submit' | 'reset' | 'button';
}

function Button(props: ButtonLinkProps | ButtonProps) {
  const { style = 'pill', variant = 'primary', ...otherProps } = props;
  const className = clsx('button', {
    [`button--${style}`]: style,
    [`button--${variant}`]: variant,
  });

  if ('to' in otherProps) {
    return <Link className={className} {...otherProps} />;
  }

  return <button className={className} {...otherProps} />;
}

export default Button;
