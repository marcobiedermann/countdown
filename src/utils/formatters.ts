function formatTime(value: number) {
  return Math.abs(value).toString().padStart(2, '0');
}

export { formatTime };
