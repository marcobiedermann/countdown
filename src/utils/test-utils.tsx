import { render as rtlRender } from '@testing-library/react';
import i18n from 'i18next';
import type { ReactNode } from 'react';
import { I18nextProvider, initReactI18next } from 'react-i18next';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import deTranslation from '../../public/locales/de/translation.json';
import enTranslation from '../../public/locales/en/translation.json';
import { options } from '../i18n';
import { store } from '../store';

i18n.use(initReactI18next).init({
  ...options,
  resources: {
    de: {
      translation: deTranslation,
    },
    en: {
      translation: enTranslation,
    },
  },
});

interface WrapperProps {
  children: ReactNode;
}

function render(ui: ReactNode, { route = '/' } = {}) {
  function Wrapper(props: WrapperProps) {
    const { children } = props;

    return (
      <Provider store={store}>
        <I18nextProvider i18n={i18n}>
          <MemoryRouter initialEntries={[route]}>{children}</MemoryRouter>
        </I18nextProvider>
      </Provider>
    );
  }

  return {
    ...rtlRender(ui, { wrapper: Wrapper }),
  };
}

export { render };
