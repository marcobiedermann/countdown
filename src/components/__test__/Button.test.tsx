import { describe, expect, it } from 'vitest';

import { render } from '../../utils/test-utils';
import Button from '../Button';

describe('Button', () => {
  it('should render the submit button correctly', () => {
    const { container } = render(<Button type="submit">Button</Button>);

    expect(container).toMatchSnapshot();
  });

  it('should render link button correctly', () => {
    const { container } = render(<Button to="#">Button</Button>);

    expect(container).toMatchSnapshot();
  });
});
