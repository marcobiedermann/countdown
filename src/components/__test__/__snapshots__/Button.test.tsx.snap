// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should render link button correctly 1`] = `
<div>
  <a
    class="button button--pill button--primary"
    href="/"
  >
    Button
  </a>
</div>
`;

exports[`Button > should render the submit button correctly 1`] = `
<div>
  <button
    class="button button--pill button--primary"
    type="submit"
  >
    Button
  </button>
</div>
`;
